Fix diverging epochs

fvaleri opened the pull request:
This patch fixes the diverging epoch problem between source and destination clusters in Cluster Mirroring. See design document for more info.

fvaleri committed:
Fix diverging epochs

This patch fixes the diverging epoch problem between source and destination clusters in Cluster Mirroring.
See design document for more info.

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

showuon commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -130,8 +130,29 @@ class ReplicaFetcherThread(name: String,
+        lastBatchEpoch = batches.next().partitionLeaderEpoch()
+      }
+      Math.max(lastBatchEpoch, partitionLeaderEpoch)
+    }
```
So this is the key fix for this issue, and others are just refactoring?

showuon replied:
I think it will be better if we add a `mirrorName.isNotEmpty` check here to apply this change to mirror topics only. WDYT?

showuon replied:
OK, it looks like the `mirrorName` in ReplicaFetcherThread is empty. Never mind.

showuon reviewed (approved):
@fvaleri , thanks for the fix! LGTM! But a high-level comment and question for your contribution to here or other places:

The key fix + comments above it might be just 20+ lines of changes. If you can make this PR as 20+ lines of changes, it will make the reviewer quickly get the change and thoughts behind it. Right now, it contains 300+ lines of changes, it took me some time to figure out all other changes are just refactoring, am I correct? 

Please try to avoid to include unrelated changes in a PR, especially it contains hundreds of lines. You can always create a separate PR to do the refactoring, so that reviewers know the refactoring PR is not changing the logic at all, which will make the review easier. 

The other thing is that you didn't mention the refactoring in the PR description, which make me thought originally all 300+ lines are related to bug fix. But again, thanks for your PR!

fvaleri commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -2603,6 +2568,74 @@ class ReplicaManager(val config: KafkaConfig,
+   * @param readOnlyLeaders Map of partitions to their metadata for partitions that became
+   *                        read-only leaders on this broker
+   */
+  private def maybeCreateMirrorFetchers(readOnlyLeaders: mutable.Map[TopicPartition, LocalReplicaChanges.PartitionInfo]): Unit = {
```
This method contains a bug fix and it is another key piece (point 2 in design doc). It is not just refactoring.

showuon replied:
Sorry, I didn't get this. Where is the fix for point 2 (Uninitialized epoch fix)? I thought the logic is the same as `applyLocalFollowersDelta` before?

fvaleri replied:
This is the fix I got partially right in the closed PR. It was hard to spot because mirror leader and follower logic were mixed up in applyLocalFollowersDelta.

Old code is skipping makeFollower for read-only leaders (see isNewLeaderEpoch when hasRemoteLeader=true). This means that partition.leaderEpoch = -1 (initial value). Then, Mirror Fetcher tries to append to the log causing UnifiedLog.analyzeAndValidateRecords to skip all batches in a tight loop as batch.partitionLeaderEpoch() > -1. This repeats for every mirror fetch request to the source cluster. The end result is 100% CPU utilization and mirror lag grows forever.

New code sets partition.leaderEpoch in applyLocalLeadersDelta, and then maybeCreateMirrorFetchers uses the correct value to start mirror fetchers.

showuon replied:
Thanks for the explanation.

showuon commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -130,8 +130,29 @@ class ReplicaFetcherThread(name: String,
+    // For regular partitions:
+    //   Batch epochs should always match the fetch state epoch. If batch epoch > fetch state epoch,
+    //   this indicates an unexpected condition. However, using the batch epoch is safe since it comes
+    //   from the leader's log, which is authoritative.
```
Sorry, on second thought, this is not correct for regular partitions. Before this change, when `batch epoch > fetch state epoch`, the batch will be ignored and exception will be thrown. So, we should only apply this effectiveLeaderEpoch to mirrored topics.

fvaleri replied:
Yeah, this is a critical point. If we use the effective epoch for regular fetches, there is a risk of log divergence because we are bypassing fetch validations protecting against log divergence (KAFKA-18723).

https://github.com/apache/kafka/blob/4.1.0/core/src/main/scala/kafka/server/AbstractFetcherThread.scala#L369-L371

Last commit adds the gating logic to apply the effective epoch to mirror partition only. We also need to set mirrorName on follower partitions.

fvaleri replied:
I think the same race also affect mirror partitions replication with effective epoch:

1. Fetch request sent at epoch X
2. Leadership changes to epoch X+1, log truncates
3. Slow fetch response arrives with batches from epoch X
4. These batches may include data that was truncated

fvaleri replied:
Last commit intoduces a new approach that is not skipping epoch validation, but simply tracking source epoch for mirror partitions. Design doc updated.

- Updated getOrCreatePartition to set mirrorName from metadata for ALL partition replicas (leaders and followers) so that we can create special logic 
- Instead of bypassing hasHigherPartitionLeaderEpoch validation with effective epoch, we update fetch epoch progressively in AbstractFetcherThread after log append

Example:

```
Source Cluster (epoch 100, 105, 110)
      |
      | consumer fetch
      V
MirrorFetcherThread (destination broker-4, read-only leader)
  - partition.mirrorName = "a2b"
  - currentLeaderEpoch: 0 -> 100 -> 105 -> 110
  - Appends with leaderEpoch from fetch state
  - Batches with epoch > currentLeaderEpoch skipped
      |
      | replica fetch
      V
ReplicaFetcherThread (destination broker-5, follower)
  - partition.mirrorName = "a2b"
  - currentLeaderEpoch: 5 -> 100 -> 105 -> 110
  - Appends with leaderEpoch from fetch state
  - Batches with epoch > currentLeaderEpoch skipped
```

showuon replied:
@fvaleri , I'm thinking about the issue in https://issues.apache.org/jira/browse/KAFKA-18723 . After this PR, it means we cannot catch the issue in KAFKA-18723. Am I correct? So complicated...

fvaleri replied:
This is probably the most complicated challenge for async mirroring. How to reconcile source and destination epoch while keeping all append validations (log corruption and divergence protections).

I now have an aggressive test procedure with multiple leader election in source and destination clusters and current solution fails thus test. It means that we need more time to think about this and experiment.

I'll post an update when I have one.

showuon replied:
I think our case should be fine.
```
/* During replication of uncommitted data it is possible for the remote replica to send record batches after it lost
 * leadership. This can happen if sending FETCH responses is slow. There is a race between sending the FETCH
 * response and the replica truncating and appending to the log. The replicating replica resolves this issue by only
 * persisting up to the current leader epoch used in the fetch request. See KAFKA-18723 for more details.
 */
skipRemainingBatches = skipRemainingBatches || hasHigherPartitionLeaderEpoch(batch, origin, leaderEpoch);
```

I read this comments many times, and I think I know understand what it is talking about, and what it tries to protect from. This is the scenario it wants to protect:

B0 (leader) contains data:
offset 0: a (leader epoch 0)
offset 1: b (leader epoch 0)
offset 2: c (leader epoch 0)

B1 contains data:
offset 0: a (leader epoch 0)


B2 contains data:
offset 0: a (leader epoch 0)
offset 1: b (leader epoch 0)

ISR: B1, B2, B3

----
When B2 fetches offset 2 from B0, we expect it should return c, but the race condition could happen:
1. B0 received the fetch request for offset 2 from B2
2. leadership change to B1
3. B0 truncate the offset to offset 0 to sync up with B1

B0 (leader) contains data:
offset 0: a (leader epoch 0)

4. producer sends a new record `d` `e` to leader B1

B1 contains data:
offset 0: a (leader epoch 0)
offset 1: d (leader epoch 1)
offset 2: e (leader epoch 1)

5. B0 fetches offset 1, 2 record from B1

B0 (leader) contains data:
offset 0: a (leader epoch 0)
offset 1: d (leader epoch 1)
offset 2: e (leader epoch 1)

6. Then, B0 sends the fetch response data to B2 in step (1), with record `e`

B2 contains data:
offset 0: a (leader epoch 0)
offset 1: b (leader epoch 0)
offset 2: e (leader epoch 1)


So in this case, B2 will contain inconsistent data and cannot be identified to be truncated. Because:
a. When B2 sends fetch request to B1, with last fetched leader epoch: 1, offset 3, the leader will think, B2 already has offset 2 with leader epoch 1, which matches what I have, so no truncation needed.
b. After the fix of `KAFKA-18723`, the step (6) cannot be appended. Then the next fetch to B1, it will send with last fetched leader epoch: 0, offset 2, and B1 found its leader epoch 0 ends at offset 0, so it will ask B2 to truncate to offset 0 and refetch.

I hope the explanation is clear.

Then, back to our case.... hmm..... hmmm... hmmm...hmm...

showuon replied:
Damn! our case still suffers from this issue! I thought we are safe.... :(

showuon replied:
I think we can consider to the `Rewrite source epochs to destination epochs` solution to fix this issue. Added comment in the design doc.

showuon replied:
Last night, I was thinking that because the mirror fetch only fetch until `highwatermark`, which means we won't fetch `uncommitted` data. That is, even if there is a leadership change, the committed records will not be truncated (because it is already committed). So actually we are safe from this [KAFKA-18723](https://issues.apache.org/jira/browse/KAFKA-18723) issue.

The only exception is the `unclean election`. If it happened, the assumption above (the committed records will not be truncated) will not be hold anymore. And the mirror fetch will also be impacted.

showuon replied:
I tend to make the `unclean election` as an exception for [KAFKA-18723](https://issues.apache.org/jira/browse/KAFKA-18723). Basically, if the source cluster has `unclean election`, the mirror fetch can still truncate and sync up with the new leader without issue. But once the rare case in KAFKA-18723 happened, we cannot prevent this issue happen.

fvaleri replied:
Do you mean current approach is safe? I'm not able to have a log convergence in my test.

All good with source mirroring, but on the destination:

1. A follower starts with a low destination epoch (e.g., 0 or 1)
2. The leader has already appended batches with high source epochs (e.g., 2, 3, 4)
3. The follower requests with its low epoch, and the leader detects "divergence" returning epoch=-1

fvaleri committed:
Gate effective epoch

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

fvaleri committed:
Replace effective epoch with epoch tracking

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

showuon commented:
1. Please make sure the compilation can be passed.
2. When the batch leader epoch > current leader epoch in mirrored topic, we'll skip them (suppose it's offset 100), and then we update the leader epoch and do another fetch. So, does the offset 100 will be fetched again, or we skip them and fetch next offset (suppose offset 101?)

fvaleri commented:
> 1. Please make sure the compilation can be passed.
 
Fixed.

> 2. When the batch leader epoch > current leader epoch in mirrored topic, we'll skip them (suppose it's offset 100), and then we update the leader epoch and do another fetch. So, does the offset 100 will be fetched again, or we skip them and fetch next offset (suppose offset 101?)

Let me double check.

fvaleri commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -477,9 +486,19 @@ abstract class AbstractFetcherThread(name: String,
+                          // For regular partitions, fetch epoch comes from metadata and should not be
+                          // updated from batch epochs.
+                          val updatedLeaderEpoch = if (shouldUpdateEpochFromBatches(topicPartition) && logAppendInfo.lastLeaderEpoch.isPresent) {
+                            Math.max(currentFetchState.currentLeaderEpoch, logAppendInfo.lastLeaderEpoch.get)
```
@showuon here we use last appended epoch for mirror partitions. This means that it works like normal replication, retrying skipped batches.

showuon replied:
It's not about the `last appended epoch`, it's about the next offset to fetch. Could you point me where we set the correct offset to retry?

fvaleri replied:
Sure, it's just few lines above (L481) and still driven by logAppendInfo.

```sh
val nextOffset = if (validBytes > 0) logAppendInfo.lastOffset + 1 else currentFetchState.fetchOffset
```

showuon replied:
Yes, that's the correct answer. Thanks.

showuon reviewed (approved):
LGTM!

fvaleri commented:
I'm doing some more tests with this to catch any corner case.

fvaleri committed:
Add mirror epoch discovery

The mirrorLeaderEpoch field is added to track source cluster epochs separately
from the destination cluster's currentLeaderEpoch. This allows mirrored partitions
to validate batches against the source cluster's epoch progression rather than the
destination cluster's epoch, which is critical for proper cluster mirroring when
epochs diverge between clusters.

Fix ReplicaFetcherTread using consumer fetch instead of replica fetch.

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

fvaleri committed:
Address comments

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

showuon commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -469,10 +480,29 @@ abstract class AbstractFetcherThread(name: String,
+                          firstBatchEpoch
+                        } else {
+                          currentFetchState.mirrorLeaderEpoch
+                        }
```
This is not right. We should always use the `currentFetchState.mirrorLeaderEpoch`, otherwise, the race condition in [KAFKA-18723](https://issues.apache.org/jira/browse/KAFKA-18723) will happen. That is, we need to have an `indicator` to let the follower know if the fetch response should be rejected. If you update it here, then we don't have the indicator to catch the unexpected batches.

fvaleri replied:
Right. I think that was from an early attempt that I forgot to remove.

showuon commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -649,13 +680,14 @@ abstract class AbstractFetcherThread(name: String,
       } else latestEpoch(tp)
       val state = if (lastFetchedEpoch.isPresent) ReplicaState.FETCHING else ReplicaState.TRUNCATING
 
-      // if it's readonly, always assume it's 0 so that we can get the current leader epoch from fetch response
-      val leaderEpoch = if (initialFetchState.mirrorName.nonEmpty) 0 else initialFetchState.currentLeaderEpoch
-      new PartitionFetchState(initialFetchState.topicId.toJava, initialFetchState.initOffset, Optional.empty(), leaderEpoch,
-        state, lastFetchedEpoch, initialFetchState.mirrorName.nonEmpty, 0)
```
The `0` leaderEpoch is needed to get the current leader epoch in the source cluster for `the leader in the target cluster`. But for `the follower in the target cluster`, it needs to send with the `initialFetchState.currentLeaderEpoch`.

fvaleri replied:
Sorry, not sure what you mean here. We are sending initialFetchState.currentLeaderEpoch.

showuon replied:
My question is, when `the the leader in the target cluster` tries to build the `partitionFetchState`, it'll use this one in `AbstractFetcherThread` because we don't override in `MirrorFetcherThread`, right? If so, it's not right we send with `initialFetchState.currentLeaderEpoch` here for `the the leader in the target cluster` (should be 0).

showuon commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -649,13 +680,14 @@ abstract class AbstractFetcherThread(name: String,
-      new PartitionFetchState(initialFetchState.topicId.toJava, initialFetchState.initOffset, Optional.empty(), leaderEpoch,
-        state, lastFetchedEpoch, initialFetchState.mirrorName.nonEmpty, 0)
+      // For mirrored partition followers, initialize mirrorLeaderEpoch to 0 and wait for FENCED_LEADER_EPOCH to update it
+      val mirrorLeaderEpoch = if (initialFetchState.mirrorName.nonEmpty) 0 else -1
```
For `the leader in the target cluster`, the `mirrorLeaderEpoch` is not needed so it should be -1.

fvaleri replied:
So the rules are:

- Mirrored partition LEADER in target cluster: This replica fetches from the source cluster leader. It needs currentLeaderEpoch = 0 (to discover source epoch) and mirrorLeaderEpoch = -1 (because the leader doesn't need mirrorLeaderEpoch)

- Mirrored partition FOLLOWER in target cluster: This replica fetches from the target cluster leader (intra-cluster). It needs currentLeaderEpoch = initialFetchState.currentLeaderEpoch (target cluster's leader epoch) and mirrorLeaderEpoch = 0 (to validate batches from source)

This is what I'm implementing now.

showuon replied:
Yes, this is what I expect. Thanks.

showuon commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -1449,6 +1449,20 @@ class Partition(val topicPartition: TopicPartition,
+            // Follower's mirrorLeaderEpoch is outdated
+            throw new FencedLeaderEpochException(
+              s"Follower's mirrorLeaderEpoch $followerMirrorEpoch is lower than leader's highest epoch $logHighestEpoch for mirrored partition $topicPartition"
+            )
```
This matches my initial idea. It works well, but has a problem:
While a leader in target cluster is mirroring from the source cluster, the `logHighestEpoch` will change on new records fetched. For example, current `logHighestEpoch=10` and we send back to follower, but the logHighestEpoch might become 11 in next fetch from source cluster, and 12, in next next fetch, and 13, 14, ... If we use this `FencedLeaderEpochException` way, the follower in the target cluster will keep getting the exception and cannot fetch the records successfully, and never catch up with the leader.

In the example in the [doc](https://docs.google.com/document/d/1_KkzgnKsGUjZ3g2pG1ckXaGLRJJJP9NOi6LD_XnCSWg/edit?disco=AAABuB7Gb08), I actually updated to the new solution without FENCED_LEADER_EPOCH error. But sorry, I didn't update the description in the doc above. So in short, my thought is:
1. The leader return batches as usual without checking the `mirrorLeaderEpoch`. So we can make sure the follower can continue to fetch the batches in the fetch request, even if the leader keeps updating the `mirrorLeaderEpoch` value.
2. Before leader sends the fetch response, we need to add the `mirrorLeaderEpoch` in the fetch response when the `mirrorLeaderEpoch` in the fetch request is not -1. 
3. Because of (2), maybe it makes more sense we move the `MirrorLeaderEpoch` field one layer above as updated in the [doc](https://docs.google.com/document/d/1_KkzgnKsGUjZ3g2pG1ckXaGLRJJJP9NOi6LD_XnCSWg/edit?disco=AAABuB7Gb2U).
4. When the follower receives the fetch response, it'll `processPartitionData` as usual. This way we can make sure the race condition in KAFKA-18723 can be identified using the `MirrorLeaderEpoch` in the fetch request. But after the process, we should update the `MirrorLeaderEpoch` in the follower to the updated value in the fetch response. So that the follower can send with the updated value in next fetch request.

Do you think this update make sense?

fvaleri replied:
Totally make sense, in fact I was not convinced we needed FENCED_LEADER_EPOCH as we are already updating the mirrorLeaderEpoch after appending to the log, which is at the core of my initial solution. The strategy you describe is almost there, I just need to remove the code highlighted and move some code for point 2.

showuon reviewed (commented):
Thanks for making the commit isolated. It makes the reviewer much easier. But I cannot just review the single commit and approve whole PR. Like you keep updating the PR after this commit, and it's hard to review commit by commit. I know you want to base on the refactor branch to develop. What you can do is:
1. Extract the refactor part into a separate PR. This should be done first. So that you don't need to mix everything in one PR.
2. While the refactor PR is opened, you can based on that PR and open another PR against it, like what I did [here](https://github.com/showuon/kafka/pull/34). I'm merging it to `deleteMirrorAttach` branch, not `cl3`. So in your case, after the refactor PR merged, you can rebase with the cl3 and continue the development. And for the reveiwers, they can review a PR with only the required changes without other noises. 

Thank you.

fvaleri commented:
@showuon all your comments should be addressed and design doc updated. Don't worry about refactorings, I've put them aside and will reapply once this is merged.

showuon commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -1449,6 +1449,20 @@ class Partition(val topicPartition: TopicPartition,
+            info(s"==== Fencing follower: partition=$topicPartition, followerMirrorEpoch=$followerMirrorEpoch, logHighestEpoch=$logHighestEpoch")
+            // Follower's mirrorLeaderEpoch is outdated
+            throw new FencedLeaderEpochException(
+              s"Follower's mirrorLeaderEpoch $followerMirrorEpoch is lower than leader's highest epoch $logHighestEpoch for mirrored partition $topicPartition"
```
I think this change is not needed.

fvaleri replied:
I think it is needed for liveness, not for correctness.

This is what happens without leader-side fencing:

1. Follower skips batches with higher epochs (correct, no divergence)
2. Follower never updates mirrorLeaderEpoch (stuck)
3. Infinite loop of fetch -> skip -> retry (no progress)

This the exact problem I was stuck with after my initial solution.

showuon replied:
Why does the follower never updates `mirrorLeaderEpoch` in the step (2)? In the current logic, we expect `AbstractFetcherThread#processFetchRequest` should do:
1. processPartitionData
2. Skip the batches with higher epochs
3. Update the mirrorLeaderEpoch by the fetch response from the leader
4. Retry the skipped offset

So, could you explain to me why does followers never update mirrorLeaderEpoch?

fvaleri replied:
They should, but practice doesn't agree. Le me debug my code.

showuon replied:
Yes, we should figure it out, instead of workaround it without knowing the reason. Otherwise, the whole solution in the PR might base on a wrong understanding about how follower works.

fvaleri replied:
Thanks for help. Last commit should work fine.

showuon commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -435,10 +480,15 @@ abstract class AbstractFetcherThread(name: String,
+                        currentFetchState.mirrorLeaderEpoch
+                      } else {
+                        currentFetchState.currentLeaderEpoch
+                      }
```
With this change, the "leader in the target cluster" will also use the `mirrorLeaderEpoch`, which is not expected. Am I right?

fvaleri replied:
Why is it not expected? Both leader and follower on the destination cluster need to validate batches using "source" epoch from fetch state, which is hosted in mirrorLeaderEpoch.

showuon replied:
As [your comment](https://github.com/showuon/kafka/pull/41#discussion_r2542192813) above: 

> So the rules are:
>   -  Mirrored partition LEADER in target cluster: This replica fetches from the source cluster leader. It needs currentLeaderEpoch = 0 (to discover source epoch) and mirrorLeaderEpoch = -1 (because the leader doesn't need mirrorLeaderEpoch)

In the mirrored partition leader, we don't need the `mirrorLeaderEpoch`. The `currentLeaderEpoch` in the  `currentFetchState` will be updated to the source epoch in `updateMirrorCachedLeaderEpoch` [here](https://github.com/showuon/kafka/pull/41/files#diff-508e9dc4d52744119dda36d69ce63a1901abfd3080ca72fc4554250b7e9f5242R287). So we should directly use `currentFetchState.currentLeaderEpoch` for mirrored partition leader.

fvaleri replied:
I see now. Thanks.

showuon commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -452,9 +502,20 @@ abstract class AbstractFetcherThread(name: String,
+                            Math.max(currentFetchState.mirrorLeaderEpoch, logAppendInfo.lastLeaderEpoch.get)
+                          } else {
+                            currentFetchState.mirrorLeaderEpoch
+                          }
```
1. Is the `logAppendInfo.lastLeaderEpoch.isPresent` needed? What is that for?
2. I don't get this: `Math.max(currentFetchState.mirrorLeaderEpoch, logAppendInfo.lastLeaderEpoch.get)`. In which case we want to use the `logAppendInfo.lastLeaderEpoch`?
3. We should adopt the `mirrorLeaderEpoch` responded from the fetch response, not from the `currentFetchState`.

fvaleri replied:
You are correct. Updated.

showuon commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -452,9 +502,20 @@ abstract class AbstractFetcherThread(name: String,
+                          // This allows source cluster epochs to be progressively accepted while maintaining
+                          // epoch validations (batches with epochs higher than fetch epoch are skipped).
+                          // For regular partitions, fetch epoch comes from metadata and should not be
+                          // updated from batch epochs.
```
This comment should be updated.

fvaleri replied:
Sure.

showuon commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -110,6 +110,29 @@ class ReplicaFetcherThread(name: String,
+    }
+  }
+
+  override protected def partitionFetchState(tp: TopicPartition, initialFetchState: InitialFetchState, currentState: PartitionFetchState): PartitionFetchState = {
```
If we override it, we should also handle some error cases like AbstractFetcherThread did.

showuon commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -2603,6 +2577,74 @@ class ReplicaManager(val config: KafkaConfig,
+              val fetchState = InitialFetchState(
+                topicId = Some(info.topicId),
+                leader = leaderEndpoint,
+                currentLeaderEpoch = partition.getLeaderEpoch,
```
Maybe we can directly set to 0 here, so that we don't need to worry about `partitionFetchState` override?

showuon commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -2542,52 +2533,35 @@ class ReplicaManager(val config: KafkaConfig,
       // Stopping the fetchers must be done first in order to initialize the fetch
       // position correctly.
       replicaFetcherManager.removeFetcherForPartitions(partitionsToStartFetching.keySet)
-      mirrorFetcherManager.removeFetcherForPartitions(partitionsToStartFetching.keySet)
```
We should keep this because the partition might be the leader in target cluster (using `mirrorFetcherManager`), and not change to follower.

fvaleri replied:
If it's the leader, then maybeCreateMirrorFetchers is called which execute replicaFetcherManager.removeFetcherForPartitions(localLeaders.keySet).

showuon replied:
Yes, but I mean:
B0 was the leader in the target cluster, so it creates a fetcher using `mirrorFetcherManager`. When it becomes the follower, we should remove the fetcher in `mirrorFetcherManager`. Change from leader to follower in the target cluster will not call `maybeCreateMirrorFetchers`.

fvaleri replied:
Ok.

showuon commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -2603,6 +2577,74 @@ class ReplicaManager(val config: KafkaConfig,
+   * @param readOnlyLeaders Map of partitions to their metadata for partitions that became
+   *                        read-only leaders on this broker
+   */
+  private def maybeCreateMirrorFetchers(readOnlyLeaders: mutable.Map[TopicPartition, LocalReplicaChanges.PartitionInfo]): Unit = {
```
Same for the other comment. If we want to create another method for mirror fetchers, we should handle the error cases like originally did in `applyLocalFollowersDelta`, like controlled shutdown case, new leader epoch case, ... etc.

showuon replied:
Let's add a `TODO` on the `maybeCreateMirrorFetchers` method to mention we should also handle the error cases, and do it later.

showuon commented on core/src/main/scala/kafka/server/mirror/MirrorFetcherThread.scala:
```diff
@@ -50,22 +57,176 @@ class MirrorFetcherThread(name: String,
+    true
+  }
+
+  override def processPartitionData(
```
1. Is this override needed now? 
2. If it's not needed, maybe we can remove the changes in the `MirrorFetcherThread` and directly `extends ReplicaFetcherThread` like before?

fvaleri replied:
I think having them separated is much more clean in terms of long term maintenance at the cost of small duplication.

fvaleri committed:
Address comments

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

fvaleri commented:
@showuon addressed few comments, still working on the others.

showuon commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -502,15 +502,11 @@ abstract class AbstractFetcherThread(name: String,
+                            fetchPartitionData.mirrorLeaderEpoch.get().intValue()
                           } else {
-                            currentFetchState.mirrorLeaderEpoch
+                            currentFetchState.currentLeaderEpoch
```
For normal followers, we expect the `mirrorLeaderEpoch` to be -1, right? If we set it as `currentLeaderEpoch`, it'll be >=0, and we will treat it as a valid value, which is not expected.

fvaleri replied:
Ok.

fvaleri committed:
Address comments

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

showuon commented on core/src/main/scala/kafka/server/AbstractFetcherThread.scala:
```diff
@@ -448,13 +492,19 @@ abstract class AbstractFetcherThread(name: String,
                         val lag = Math.max(0L, partitionData.highWatermark - nextOffset)
                         fetcherLagStats.getAndMaybePut(topicPartition).lag = lag
 
+                        val shouldUpdateState = (validBytes == 0 && partitionData.mirrorLeaderEpoch > currentFetchState.mirrorLeaderEpoch) ||
+                          (validBytes > 0 || currentFetchState.lag.isEmpty)
```
Let's have some comment here to explain what this check is for: `(validBytes == 0 && partitionData.mirrorLeaderEpoch > currentFetchState.mirrorLeaderEpoch)`

fvaleri replied:
Sure.

showuon commented on core/src/main/scala/kafka/server/ReplicaFetcherManager.scala:
```diff
@@ -37,7 +37,7 @@ class ReplicaFetcherManager(brokerConfig: KafkaConfig,
   override def createFetcherThread(fetcherId: Int, sourceBroker: BrokerEndPoint): ReplicaFetcherThread = {
     val threadName = s"ReplicaFetcherThread-$fetcherId-${sourceBroker.id}"
-    val logContext = new LogContext(s"[ReplicaFetcher replicaId=${brokerConfig.brokerId}, leaderId=${sourceBroker.id}, fetcherId=$fetcherId]")
+    val logContext = new LogContext(s"[ReplicaFetcher replicaId=${brokerConfig.brokerId}, leaderId=${sourceBroker.id}, fetcherId=$fetcherId] ")
```
nit: Unnecessary change.

fvaleri replied:
Ok. Will do this in a follow up refactoring PR.

showuon commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -2549,45 +2540,29 @@ class ReplicaManager(val config: KafkaConfig,
           case Some(node) =>
             val log = partition.localLogOrException
+            // Pass mirrorName to InitialFetchState so the fetcher knows to use source epoch logic
+            val mirrorName = if (partition.mirrorName != null && partition.mirrorName.nonEmpty) partition.mirrorName else ""
```
nit: We should only expect `normal followers` enter `applyLocalFollowersDelta` now, so the mirror name should always be "", right?

fvaleri replied:
This flag is used to distinguish mirror from regular followers. It is used in partitionFetchState method and triggers different epoch handling logic throughout the fetch lifecycle.

showuon replied:
Oh, you are right! Please ignore this then. Thanks.

showuon reviewed (approved):
LGTM! Just some minor comments. You can merge it when updated. Thanks.

fvaleri committed:
Add nits

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>
